Add appeals

Gargron opened the pull request:
Fix #9791

Allow users to view "strikes" (actions taken against an account) and to submit appeals through the form. Moderators are notified about the appeals and can choose to approve them to undo the strike, e.g. unsuspend a suspended account, or restore deleted posts. User is notified if their appeal is approved.

![image](https://user-images.githubusercontent.com/184731/153477718-9c9592e7-2e27-4ff1-ba91-02a42e4a85d1.png)
![image](https://user-images.githubusercontent.com/184731/153477973-b07f359b-bd2d-4f50-b1b2-3b7c3ee2696e.png)

___

Funded through the BMBF (BMBF Förderkennzeichen: 01IS21S29)

Gargron committed:
Add appeals

ClearlyClaire commented:
I don't understand why you don't want to put the username right besides the display name like everywhere else here
![image](https://user-images.githubusercontent.com/384364/153492003-223e2f37-40b1-4cb7-8f2c-835f37acc3a3.png)

ClearlyClaire commented:
You can only approve appeals? … you can't reject them?

Gargron commented:
To reject them you just do nothing

ClearlyClaire commented:
Does this mean that if no moderator has report email enabled (which is probably a bad thing), moderators can't see whether there are appeals?

Gargron commented:
Yes, it does mean that

ClearlyClaire commented:
ok, that's probably a shortcoming then…

Gargron commented:
Seems like it could be added later

ClearlyClaire commented:
I'd rather be sure that a new moderation feature is complete and doesn't hinge on mail delivery being enabled and reliable for every involved party.

Gargron committed:
Add ability to reject appeals and ability to browse pending appeals in admin UI

Gargron commented:
Browsing appeals:

![image](https://user-images.githubusercontent.com/184731/153733735-e3a403ba-786f-457d-9b7f-ae5bd7d1ea71.png)

Visible on dashboard:

![image](https://user-images.githubusercontent.com/184731/153733741-3c8ba1aa-5dda-4026-8f27-1007e4fb9694.png)

And on account pages:

![image](https://user-images.githubusercontent.com/184731/153733761-267bb093-8ad5-4627-b62d-88b65677efe5.png)

Gargron committed:
Add strikes to account page in settings

Gargron commented:
Strikes and appeals accessible on account settings page:

![image](https://user-images.githubusercontent.com/184731/153738551-fe8f45e4-a69b-4529-af32-4793fcea1ccb.png)

ClearlyClaire commented on app/views/disputes/strikes/show.html.haml:
```diff
@@ -0,0 +1,127 @@
+        .report-header__details__item__header
+          %strong= t('disputes.strikes.associated_report')
+        .report-header__details__item__content
+          = link_to t('admin.reports.report', id: @strike.report.id), admin_report_path(@strike.report)
```
The styling for this one seems wrong:
![image](https://user-images.githubusercontent.com/384364/153853440-c50bf21c-4215-4b44-9502-c501ca3c3bc2.png)

ClearlyClaire reviewed (changes requested):
First, thank you for addressing my previous concerns, namely:
- the display name being used without the username alongside (now, only the username is displayed), which could have been used for phishing-type attacks
- appeals not showing up anywhere besides in mail notifications (they can now be accessed from any place I would expect, and can be listed by moderators)

In addition to the inline comments, I think the feature can be useful and at least worth considering, **however** I'm worried about the signal-to-noise ratio and the amount of junk and abusive appeals admins will get from it. In that sense, I would prefer if:
- admins could opt out of the feature altogether (without making users think they can appeal) if it turns out to be too much trouble
- moderators could chose to not be notified of appeals while still notified of reports (so that work can be split differently etc.)

In addition, I have other concerns:
- **reverting post deletion will just cause issues** as by that time, post deletions are already federated (and it's a good thing for that to be near-immediate, as remote reporters would not understand the posts not being deleted), which is, to most implementations, irreversible (in order to prevent replay attacks etc.)
- I don't know if it's a good idea to notify users of rejected appeals
- I don't think it's good to list old rejected strikes in an otherwise-unrelated page in the account settings (for actionable strikes, it's a good thing); maybe strikes should even expire after a long while (even from moderator view) but I'm really not sure about that
- overall, I don't really know about the timing of the appeal process: I think replying immediately can be useful for obvious moderation errors, but detrimental when the dispute is founded; it also may make sense for someone to apologize and show improvements months later; but at the same time, some things, like suspensions, are irreversible by that time

Gargron commented on app/controllers/admin/dashboard_controller.rb:
```diff
@@ -8,6 +8,7 @@ def index
       @pending_users_count   = User.pending.count
       @pending_reports_count = Report.unresolved.count
       @pending_tags_count    = Tag.pending_review.count
+      @pending_appeals_count = Appeal.pending.count
```
If the table grows large enough. This is a page that is only visited by a few people so saving an index seems fine here.

ClearlyClaire replied:
This page is only visited by a few people but you want it to be fast.

Gargron replied:
I insist on delaying it until at least a separate PR due to difficulties maintaining a clean schema.rb between PRs.

Gargron commented on app/controllers/admin/disputes/appeals_controller.rb:
```diff
@@ -0,0 +1,40 @@
+    authorize @appeal, :approve?
+    log_action :reject, @appeal
+    @appeal.reject!(current_account)
+    UserMailer.appeal_rejected(@appeal.account.user, @appeal)
```
I think that if we have an explicit reject function, we might as well be upfront and notify people. Imagine you are appealing a copyright strike on YouTube and they wouldn't notify you about the final decision on that.

Gargron commented on app/controllers/admin/disputes/appeals_controller.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  def filter_params
+    params.slice(:page, *Admin::AppealFilter::KEYS).permit(:page, *Admin::AppealFilter::KEYS)
```
Just for consistency with other `filter_param` methods in other controllers.

Gargron commented on app/controllers/auth/registrations_controller.rb:
```diff
@@ -123,6 +126,10 @@ def set_sessions
     @sessions = current_user.session_activations
   end
 
+  def set_strikes
+    @strikes = current_account.strikes.where(appealed_at: nil).latest
+  end
```
What do you call it when a strike has been successfully appealed? Is it "overruled"?

ClearlyClaire replied:
How is this solved?

Gargron commented on app/services/approve_appeal_service.rb:
```diff
@@ -0,0 +1,74 @@
+    target_account.user.enable!
+  end
+
+  def undo_delete_statuses!
+    @strike.statuses.each(&:undiscard)
+  end
```
I misread something in the code, you are correct.

ClearlyClaire commented on app/controllers/admin/accounts_controller.rb:
```diff
@@ -146,7 +146,7 @@ def filtered_accounts
 
     def filter_params
-      params.slice(*AccountFilter::KEYS).permit(*AccountFilter::KEYS)
+      params.slice(:page, *AccountFilter::KEYS).permit(:page, *AccountFilter::KEYS)
```
To be honest it's a mess currently, and I'm not sure having them in `filter_params` helps, you seem to need more code to filter it out, and you never access it through `filter_params`. I think I'd rather change the other controllers in a different way. But in another PR!

Gargron replied:
Actually, realized this is needed if we want batch actions to redirect us back to the same page. So it was a bug before.

Gargron commented on app/helpers/admin/trends/statuses_helper.rb:
```diff
@@ -0,0 +1,20 @@
+
+    return '' if text.blank?
+
+    html = Formatter.instance.send(:encode, text)
+    html = Formatter.instance.send(:encode_custom_emojis, html, status.emojis, prefers_autoplay?)
```
Uncomfortable making that step now. Ideally Formatter needs a rewrite where it no longer relies on a status object being passed to it.

Gargron committed:
Various fixes and improvements

- Add separate notification setting for appeals, separate from reports
- Fix style of links in report/strike header
- Change approving an appeal to not restore statuses (due to federation complexities)
- Change style of successfully appealed strikes on account settings page
- Change account settings page to only show unappealed or recently appealed strikes

Gargron committed:
Change appealed_at to overruled_at

ClearlyClaire commented on app/views/auth/registrations/_status.html.haml:
```diff
@@ -1,22 +1,17 @@
-      = link_to t('migrations.cancel'), settings_migration_path
-    - else
-      %span.positive-hint= t('auth.status.functional')
+= render partial: 'account_warning', collection: @strikes
```
Better now, but I wouldn't want to encourage someone to appeal a strike to make it disappear. It also doesn't make sense when the user cannot appeal to the strike anymore. The check should probably be based on `created_at`.

ClearlyClaire replied:
I guess it still does, but what about strikes for deleted content?

Gargron replied:
I'm still of the opinion that past warnings should be accessible. And even though you may be unsilenced separately, this is still history that happened.

Gargron committed:
Fix missing method error

nightpool commented:
I think being able to reverse post deletions is an important moderation tool, and I would prefer that we do the work to support that (although for scope reasons it probably should be separate from this PR? so I'm okay with having undeletions be slightly weird or broken until then)

nightpool commented:
that seems fine to me as an interim step as long as we have a plan to
address it in the future.

On Mon, Feb 14, 2022, 12:18 PM Claire ***@***.***> wrote:

> I think being able to reverse post deletions is an important moderation
> tool, and I would prefer that we do the work to support that (although for
> scope reasons it probably should be separate from this PR? so I'm okay with
> having undeletions be slightly weird or broken until then)
>
> By “sligtly weird” it means it would only work locally, and every remote
> instance that knew the posts would not be able to get it again.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastodon/mastodon/pull/17364#issuecomment-1039405231>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZCV4ELEJFXO7WY2XO3U3U3FBQTANCNFSM5MWVENVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Gargron commented:
So I should undo my undo of the undo feature?

ClearlyClaire commented:
I'd rather the interim step being the undo not being available until we can federate it.

Gargron commented:
I think I addressed everything here for now.

ClearlyClaire commented:
I'm still unconvinced by the “show any strike forever unless it has been explicitly overruled” thing. Like I said, you probably don't want to be reminded forever that your moderators deleted 3 times 2 toots of you months ago.

Otherwise, maybe make it explicit that deleted toots are not restored when the decision is overruled, until we have a solution for this.

Finally, I still think admins should be able to disable the feature entirely, but mods being able to opt out of appeal emails is already a good step, and I guess we can first see how it turns out on real instances.

Gargron commented:
This can be adjusted later, even before next release. I’m sure once i run it live I’ll find things to improve. I don’t think this should prevent a merge.

Gargron commented:
Thank you!
